### PR TITLE
Add profile to SDK

### PR DESCRIPTION
**Pull request by lmazuel**
Usage

```python
from azure.profiles import KnownProfile
from azure.mgmt.network import NetworkManagementClient

# Create a public Azure target client
client = NetworkManagementClient(creds, sub, profile=KnownProfile.latest)
client.network_interfaces.get()  # Will call floating latest, depending of the package

# Create a 2017-03-09-profile client
client = NetworkManagementClient(creds, sub, profile=KnownProfile.v2017_03_09_profile)
client.network_interfaces.get()  # Will call "2015-06-15" forever

# Create a 2017-03-09-profile client too, from an official label
client = NetworkManagementClient(creds, sub, profile=KnownProfile.from_name("2017-03-09-profile"))
client.network_interfaces.get()  # Will call "2015-06-15" forever

# Create a client that use "default"
client = NetworkManagementClient(creds, sub, profile=KnownProfile.default)
client.network_interfaces.get()  # By default, call floating latest, depending of the package

# Change default. This shift all clients, and does not require creating again the clients
KnownProfile.default.use(KnownProfile.2017_03_09_profile)

client.network_interfaces.get()  # Will call "2015-06-15", no client re-create

# Cannot set api_version and profile at the same time
# https://github.com/Azure/azure-sdk-for-python/issues/1864
try:
   NetworkManagementClient(creds, sub, api_version="2017-11-01", profile=KnownProfile.latest)
except ValueError:
   pass 

# Setting only api_version, change everything
# Warning: if RP is not consistent (VM and disks does not use same api version), this CANNOT work.
# https://github.com/Azure/azure-sdk-for-python/issues/1864
client = NetworkManagementClient(creds, sub, api_version="2017-11-01")
client.network_interfaces.get()  # Will call "2017-11-01"
client.subnets.get()  # Will call "2017-11-01"

# Still accept profile as a dict as today
client = NetworkManagementClient(creds, sub, profile={
  "network_interfaces": "2017-11-01"
})
client.network_interfaces.get()  # Will call "2017-11-01"

# However, this profile didn't define a default
try:
   client.subnets.get()
except ValueError:
   pass # No default and no value for subnet
```

### CODE REVIEW

File: azure-common/azure/profiles/__init__.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    """Store a default profile.
+
+    :var ProfileDefinition profile: The default profile as class attribute
+    """
+    profile = None
+
+    def use(self, profile):
+        """Define a new default profile."""
+        if not isinstance(profile, KnownProfiles) and not isinstance(profile, ProfileDefinition):
+            raise ValueError("Can only set as default a ProfileDefinition or a KwownProfiles")
```
Comment (derekbekoe) [inline azure-common/azure/profiles/__init__.py]: nit: typo 'KwownProfiles'

File: azure-common/azure/profiles/__init__.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+class DefaultProfile(object):
+    """Store a default profile.
+
+    :var ProfileDefinition profile: The default profile as class attribute
+    """
+    profile = None
+
+    def use(self, profile):
+        """Define a new default profile."""
+        if not isinstance(profile, KnownProfiles) and not isinstance(profile, ProfileDefinition):
```
Comment (derekbekoe) [inline azure-common/azure/profiles/__init__.py]: `isinstance` can take a tuple of classes.
May make it simpler?

File: azure-common/azure/profiles/__init__.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                None: "2015-01-01"
+            },
+            "azure.mgmt.resource.links.ManagementLinkClient": {
+                None: "2016-09-01"
+            },
+            "azure.mgmt.resource.resources.ResourceManagementClient": {
+                None: "2016-02-01"
+            },
+            "azure.mgmt.resource.subscriptions.SubscriptionClient": {
+                None: "2016-06-01"
```
Comment (derekbekoe) [inline azure-common/azure/profiles/__init__.py]: What's the `None` key for?
I'm not familiar.

File: azure-common/azure/profiles/multiapiclient.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        if api_version and profile is not KnownProfiles.default:
+            raise ValueError("Cannot use api-version and profile parameters at the same time")
+
+        if api_version:
+            self.profile = ProfileDefinition({
+                self._PROFILE_TAG: {
+                    None: api_version
+                }},
+                self._PROFILE_TAG + " " + api_version
```
Comment (derekbekoe) [inline azure-common/azure/profiles/multiapiclient.py]: Is this just so you can set a default value?

File: azure-common/azure/profiles/multiapiclient.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            )
+        elif isinstance(profile, dict):
+            self.profile = ProfileDefinition({
+                self._PROFILE_TAG: profile,
+                },
+                self._PROFILE_TAG + " dict"
```
Comment (derekbekoe) [inline azure-common/azure/profiles/multiapiclient.py]: Is this just so you can set a default value?

**Comment by derekbekoe** [comment]
```python
# Create a 2017-03-09-profile client too, from an official label
client = NetworkManagementClient(creds, sub, profile=KnownProfile("2017-03-09-profile"))
client.network_interfaces.get()  # Will call "2015-06-15" forever
```
Would this not be `KnownProfile.from_name(...)` instead?

File: azure-common/azure/profiles/__init__.py (hunk 2 above)
Comment (lmazuel) [reply azure-common/azure/profiles/__init__.py]: Yes indeed

File: azure-common/azure/profiles/__init__.py (hunk 3 above)
Comment (lmazuel) [reply azure-common/azure/profiles/__init__.py]: This means default value. Using `None` allows to be sure we have no conflict, and there is only one value possible.

File: azure-common/azure/profiles/multiapiclient.py (hunk 4 above)
Comment (lmazuel) [reply azure-common/azure/profiles/multiapiclient.py]: Yes, api_version is supported for a long time and I don't want to break. This means exactly "use this api version for everything" and then means this is the absolute default. Currently, it's the `None` key.

File: azure-common/azure/profiles/multiapiclient.py (hunk 5 above)
Comment (lmazuel) [reply azure-common/azure/profiles/multiapiclient.py]: I already support a dict as parameter, where key are operation group and value are api-version. I migrate that dict to the new internal format.

**Pull request merged**